Create a generic archive download manager

naoey opened the pull request:
Moves the beatmap download logic out of `BeatmapManager` into a base class that can be shared by other model managers that need to implement download functionality.

As per the discussion in #4985 this is the first of multiple PRs to centralise model downloading logic.

naoey committed:
Create a generic base archive download manager class

naoey committed:
Make BeatmapManager inherit from new base class

naoey committed:
Add license headers and xmldoc

naoey committed:
Add ability to perform a download request with options

naoey committed:
Remove unused usings

naoey committed:
Add method to check if model exists locally already

naoey committed:
Remove unecessary third generic and change usages to match

naoey committed:
Fix order

naoey committed:
Create interfaces for DownloadTrackingComposite to consume

naoey committed:
Rename interface

- Fix wrong inheritance in ArchiveModelManager
- Add license headers

naoey committed:
Fix doc move derp

naoey committed:
Merge remote-tracking branch 'upstream/master' into generic-download-model-manager

peppy committed:
Rename to DownloadableArchiveModelManager

peppy committed:
Rename request type to be less verbose

peppy commented on osu.Game/Database/ArchiveModelManager.cs:
```diff
@@ -44,7 +44,7 @@ public abstract class ArchiveModelManager<TModel, TFileModel> : ICanAcceptFiles
         /// This is not guaranteed to run on the update thread.
         /// </summary>
-        public event ItemAddedDelegate ItemAdded;
+        public event Action<TModel, bool> ItemAdded;
```
intentional change? if so, `ItemAddedDelegate` should be removed.

peppy replied:
still not sure about this one. the reason it had a delegate is that the bool here is ambiguous.

naoey replied:
there wasn't really any particular reason for changing this other than avoiding declaring a delegate in `IModelManager`. should it be added back on the interface or should we just add it to the documentation?

peppy commented on osu.Game/Database/IModelManager.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace osu.Game.Database
+{
+    public interface IModelManager<TModel>
```
needs xmldoc. also this doesn't contain the correct constraints on `TModel`.

also this isn't really used anywhere. not sure we should be defining such interfaces too early.

naoey replied:
agree that it was early to include in this PR. it is used in the `DownloadTrackingComposite` update PR (#5001) to consume the items added/removed events.

also was (am still) unsure what the right constraints to have here at the interface level would be.

peppy replied:
Ah, I didn't realise that uses the interface. Is that for DI purposes?

naoey replied:
well its not used directly but by virtue of `IModelDownloader` inheriting it (which is used for the DI). the `ItemAdded`/`ItemRemoved` events are needed on the generic type.

peppy replied:
Right. Understood.

peppy commented on osu.Game/Database/IModelDownloader.cs:
```diff
@@ -0,0 +1,48 @@
+
+namespace osu.Game.Database
+{
+    public interface IModelDownloader<TModel> : IModelManager<TModel>
```
(class) xmldoc should be placed on interface instead of abstract class, if we decide to go in this direction.

naoey committed:
Merge remote-tracking branch 'upstream/master' into generic-download-model-manager

naoey committed:
Merge remote-tracking branch 'upstream/master' into generic-download-model-manager

naoey committed:
Remove unused delegate, use model name in notifications, add more xmldoc

- Applies a `class` constraint to the generic type in `IModelManager`
- Add xmldoc

peppy commented on osu.Game/Online/API/ArchiveDownloadRequest.cs:
```diff
@@ -0,0 +1,24 @@
+    public abstract class ArchiveDownloadRequest<TModel> : APIDownloadRequest
+        where TModel : class
+    {
+        public readonly TModel Info;
```
Rename this to Model

naoey committed:
Rename Info to Model

naoey committed:
Move doc to interface

peppy committed:
Merge branch 'master' into generic-download-model-manager

naoey committed:
Remove extra newline

peppy committed:
Merge branch 'master' into generic-download-model-manager

peppy committed:
Move private event handling logic to bottom of class

peppy committed:
Simplify download method

peppy commented:
I made some changes to simplify the download logic (since we don't really use non-typed objects anywhere in osu!). Please check you agree with this and also that it works OK for your future PRs.

peppy committed:
Make generic covariant

naoey commented:
I had initially implemented something like this, but I considered that a binary value option may be too limiting for some downloads in the future. Of course this depends on systems that don’t exist right now, so I’ll leave it to you to decide which is better based on the direction you’d want it to take. Or whether to revisit this at that point. 

Eg. I was thinking that with something like a skin download, we might want to use multiple parameters like which game mode(s), whether to include extras like cursor and hitsound packs etc.

Other than that this will work as is with my next PRs 👍

peppy commented:
If we need more than just a `bool`, the correct direction would be to make a generic DownloadRequest object which can be derived:

```
interface DownloadRequest<T>
{
    T Model { get; set; }
}

BeatmapDownloadRequest : DownloadRequest<BeatmapSetInfo>
{
    BeatmapSetInfo Model { get; set; }

    public bool NoVideo { get; set; }
}
```

peppy committed:
Merge branch 'master' into generic-download-model-manager

naoey commented:
But `CreateDownloadRequest` would need the parameters to initialise the DownloadRequest object correctly?

peppy commented:
it would become:

`CreateDownloadRequest(DownloadRequest<TModel> request)`

naoey commented:
That seems a bit backwards. `CreateDownloadRequest` should take a `TModel` argument and return a `DownloadRequest<TModel>`, but here it takes the `DownloadRequest<TModel>` as the argument?

In that case can't we just expose the `performDownloadWithRequest` private method to actually do the download, and externalise the `DownloadRequest` creation completely?

peppy commented:
Oh sorry, I mistyped.

```csharp
public Download(DownloadRequest<TModel> request);

protected override ArchiveDownloadRequest<BeatmapSetInfo> CreateDownloadRequest(DownloadRequest<TModel> request);
```

The new proposed `DownloadRequest` would be the input to the download system (from the `DownloadButton` etc.). It could be made a secondary overload to allow for still making simple downlaods via `Download(TModel model)` still.

naoey commented:
Ah I see now. Yes that would be better.

peppy commented:
Best not over-complicate until necessary, so (optional) bool is fine for the time being.

naoey commented on osu.Game/Online/API/Requests/DownloadBeatmapSetRequest.cs:
```diff
@@ -2,28 +2,21 @@
-            BeatmapSet = set;
-
-            Progressed += (current, total) => DownloadProgressed?.Invoke(Progress = (float)current / total);
+            this.set = set;
```
just noticed this unnecessary local variable. `Model` is already on parent class.

peppy replied:
feel free to fix if you are able to. i'll be doing another pass on this later on failing that.

naoey committed:
Remove redundant variable, handle all request failures

peppy commented on osu.Game/Database/DownloadableArchiveModelManager.cs:
```diff
@@ -122,14 +113,27 @@ private void performDownloadWithRequest(ArchiveDownloadRequest<TModel> request)
-                catch
+                catch (Exception e)
                 {
+                    // 404s (and maybe other failures) don't fire request.Failure so for now they handled here as well
```
should be double-checked as to why before this is merged as a valid solution. (note to self if no one beats me).

peppy committed:
Merge branch 'master' into generic-download-model-manager

peppy commented:
@naoey you should be able to remove the try-catch logic after ppy/osu-framework#2572 is merged.

naoey committed:
Merge remote-tracking branch 'upstream/master' into generic-download-model-manager

naoey committed:
Remove unnecessary try-catch block

naoey committed:
Remove unnecessary private methods and inline used-once code

peppy committed:
Remove existing argument for ItemAdded event

For all usages, it looks like this was unnecessary.

peppy committed:
Merge remote-tracking branch 'upstream/master' into generic-download-model-manager

peppy commented:
I ended up removing the `bool` as it was never used correctly and does not seem necessary.

peppy committed:
Fix comments

peppy committed:
Merge branch 'master' into generic-download-model-manager
